이전 장에서 베이즈 정리를 유도했습니다:

$$
P(A∣B)= \frac{P(A)P(B∣A)}{P(B)}
$$ 

예를 들어, 우리는 General Social Survey에서 얻은 데이터를 사용하여 베이즈 정리를 사용하여 조건부 확률을 계산하는 예제를 살펴 보았습니다. 그러나 우리는 완전한 데이터 집합을 가지고 있었기 때문에 실제로는 베이즈 정리가 필요하지 않았습니다. 왼쪽 항을 직접 계산하는 것은 충분히 쉽고, 오른쪽 항을 계산하는 데는 큰 도움이 되지 않았습니다.

그러나 종종 우리는 완전한 데이터 집합을 가지고 있지 않으며, 이 경우 베이즈 정리가 더 유용합니다. 이 장에서는 조건부 확률과 관련된 여러 어려운 문제를 해결하는 데 베이즈 정리를 사용할 것입니다.

쿠키 문제로 시작하겠습니다. 이것은 굴 문제의 변형입니다:

두 개의 그릇에 쿠키가 있다고 가정해 보겠습니다.

* 그릇 1에는 바닐라 쿠키 30개와 초콜릿 쿠키 10개가 들어 있습니다.

* 그릇 2에는 바닐라 쿠키 20개와 초콜릿 쿠키 20개가 들어 있습니다.

이제 무작위로 두 그릇 중 하나를 선택하고, 보지 않고 쿠키 하나를 무작위로 선택한다고 가정해 보겠습니다. 만약 선택한 쿠키가 바닐라 쿠키라면, 이 쿠키가 그릇 1에서 나온 확률은 얼마일까요?

우리가 원하는 것은 바닐라 쿠키를 선택한 경우, 즉 $P(B_1 | V)$인 조건부 확률입니다.

그러나 문제의 진술에서 얻는 것은 다음과 같습니다:

* 그릇 1에서 선택했을 때 바닐라 쿠키를 얻을 조건부 확률, $P(V | B_1)$ 및

* 그릇 2에서 선택했을 때 바닐라 쿠키를 얻을 조건부 확률, $P(V | B_2)$.


베이즈 정리는 이들이 어떻게 관련되어 있는지를 알려줍니다:


$$ P(B_{1}|V)=\frac{P(V|B_{1})}{P(V)} $$
 

왼쪽 항은 우리가 원하는 것입니다. 오른쪽 항들은 다음과 같습니다:

$P(B_1)$: 우리가 어떤 종류의 쿠키를 얻었는지와 관계없이 Bowl 1을 선택할 확률입니다. 문제에서 우리는 무작위로 그릇을 선택했다고 가정하므로 $P(B_1) = 1/2$로 가정합니다.

$P(V|B_1)$: Bowl 1에서 바닐라 쿠키를 얻을 확률로, 3/4입니다.

$P(V)$: 어느 그릇에서든 바닐라 쿠키를 뽑을 확률입니다. 5/8

<img src='images/b1.jpg'>

<img src='images/b2.jpg'>

<a href='https://m.blog.naver.com/PostView.nhn?isHttpsRedirect=true&blogId=sw4r&logNo=221919933999'>marginal</a>

$P(V)$를 계산하기 위해 전체 확률의 법칙을 사용할 수 있습니다:
$$P(V)=(B_{1})P(V|B_{1})+(B_{2})P(V|B_{2})$$
$$ P(V)=\sum_{i=1}^{2} P(B_{i})p(V|B_{i})  $$
$$P(V)=(1/2) (3/4) + (1/2) (1/2)=5/8$$

또한 이 결과를 직접 계산할 수도 있습니다.

어떤 그릇을 선택할 확률은 같았고, 두 그릇에는 동일한 수의 쿠키가 들어 있으므로 어떤 쿠키를 선택할 확률도 동일합니다.

두 그릇에는 모두 50개의 바닐라 쿠키와 30개의 초콜릿 쿠키가 있으므로 $P(V) = 5/8$입니다.

마지막으로, 베이즈 정리를 적용하여 Bowl 1의 사후 확률을 계산할 수 있습니다:
 
$$P(B_1|V) = (1/2)~(3/4)~/~(5/8) = 3/5$$

이 예제는 베이즈 정리의 한 가지 사용법을 보여줍니다: 이것은 $P(B|A)$에서 $P(A|B)$로 넘어가는 방법을 제공합니다. 이 전략은 왼쪽 항보다 오른쪽 항을 계산하기 쉬운 경우와 같이 사용하기 유용합니다.

려하여 업데이트하는 방법을 제공합니다.

이 해석은 "다이크로닉(diachronic)"이라고도 하며, 시간이 지남에 따른 변화와 관련이 있습니다. 이 경우, 가설의 확률은 새로운 데이터를 관찰함에 따라 변화합니다.

$H$와 $D$로 베이즈 정리를 다시 작성하면 다음과 같습니다:
$$P(H|D) = \frac{P(H)~P(D|H)}{P(D)}$$
 

이 해석에서 각 항은 이름이 있습니다:

$P(H)$는 데이터를 보기 전 가설의 확률으로, 사전 확률(prior probability) 또는 간단히 **사전(prior)**이라고 합니다.

$P(H|D)$는 데이터를 보고 난 후의 가설의 확률로, **사후 확률(posterior)**이라고 합니다.

$P(D|H)$는 가설 하에서 데이터의 확률로, **우도(likelihood)**라고 합니다.

$P(D)$는 **데이터의 총 확률(total probability of the data)**입니다.

가끔은 사전 확률을 백그라운드 정보를 기반으로 계산할 수 있습니다. 예를 들어, 쿠키 문제에서는 그릇을 무작위로 선택한다는 정보를 밝히므로 사전 확률을 계산할 수 있습니다.

다른 경우 사전은 주관적일 수 있습니다. 즉, 합리적인 사람들은 다른 백그라운드 정보를 사용하거나 동일한 정보를 다르게 해석하기 때문에 의견이 분분할 수 있습니다.

우도는 일반적으로 계산하기 가장 쉬운 부분입니다. 쿠키 문제에서는 각 그릇에 들어 있는 쿠키의 수가 주어지므로 각 가설 하에서 데이터의 확률을 계산할 수 있습니다.



<a href='https://drhongdatanote.tistory.com/57'>중심극한정리</a>

<a href='https://heeya-stupidbutstudying.tistory.com/entry/%ED%86%B5%EA%B3%84-%EC%B5%9C%EB%8C%80%EC%9A%B0%EB%8F%84%EB%B2%95Maximum-Likelihood-Estimation'>MLE</a>

In [1]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt

In [2]:
table = pd.DataFrame(index=['Bowl 1','Bowl 2'])

In [3]:
table['prior']=1/2, 1/2
table['likelihood']=3/4, 1/2
table

prior  likelihood
Bowl 1    0.5        0.75
Bowl 2    0.5        0.50

여기서 이전 방법과 다른 점을 볼 수 있습니다: 우리는 Bowl 1뿐만 아니라 두 가설에 대한 우도를 계산합니다.

Bowl 1에서 바닐라 쿠키를 얻을 확률은 3/4입니다.

Bowl 2에서 바닐라 쿠키를 얻을 확률은 1/2입니다.

우리는 우도를 계산할 때 이들이 1이 되지 않는다는 것을 알 수 있습니다. 그것은 괜찮습니다. 각각은 다른 가설에 조건을 둔 확률이기 때문에 합쳐지지 않아도 괜찮으며, 그렇지 않아도 됩니다.

다음 단계는 베이즈 정리와 유사합니다. 사전 확률에 우도를 곱합니다:

In [4]:
table['unnorm']=table['prior']*table['likelihood']
table

prior  likelihood  unnorm
Bowl 1    0.5        0.75   0.375
Bowl 2    0.5        0.50   0.250

이 결과를 "정규화되지 않은 사후 확률(unnormalized posteriors)"이라고 부르는데, 이 값들은 사전과 우도의 곱으로 나타납니다:
 
이는 베이즈 정리의 분자 부분입니다. 이들을 모두 더하면 다음과 같습니다:

$$P(B_i)~P(D|B_i)$$


$$P(B_1)~P(D|B_1) + P(B_2)~P(D|B_2)$$

이것은 베이즈 정리의 분모인 $P(D)$입니다.

따라서 데이터의 총 확률을 다음과 같이 계산할 수 있습니다:



In [5]:
prob_data = table['unnorm'].sum()
prob_data

0.625

우리가 얻은 데이터는 0.625,P(D)를 직접 계산함으로써 우리가 얻게되는 것이다
그리고 우리가 이처럼 사후확률을 계산할 수 있다.

In [6]:
table['posterior'] = table['unnorm']/prob_data
table

prior  likelihood  unnorm  posterior
Bowl 1    0.5        0.75   0.375        0.6
Bowl 2    0.5        0.50   0.250        0.4

Bowl 1의 사후 확률은 0.6으로, 베이즈 정리를 명시적으로 사용하여 얻은 결과와 일치합니다. 추가로, Bowl 2의 사후 확률도 0.4로 얻습니다.

정규화되지 않은 사후 확률을 더하고 나눔으로써, 사후 확률을 1로 만듭니다. 이 과정을 "정규화(normalization)"라고 하며, 데이터의 총 확률은 "정규화 상수(normalizing constant)"로도 불립니다.

## 주사위 문제

베이스 테이블을 사용하면 두 개 이상의 가설이 있는 문제도 해결할 수 있습니다. 예를 들어 다음과 같은 문제를 생각해 봅시다:

> 6면 주사위, 8면 주사위, 12면 주사위가 들어 있는 상자가 있다고 가정해 보겠습니다. 무작위로 주사위를 선택하고 굴린 다음, 굴린 결과를 1이라고 보고했다고 가정합니다. 6면 주사위를 선택한 확률은 얼마일까요?

이 예에서는 세 개의 가설이 동일한 사전 확률을 가지고 있습니다. 데이터는 결과가 1이라고 보고한 것입니다.

6면 주사위를 선택한 경우, 데이터의 확률은 1/6입니다. 8면 주사위를 선택한 경우, 확률은 1/8이고, 12면 주사위를 선택한 경우, 확률은 1/12입니다.

다음은 가설을 정수로 나타내는 베이스 테이블입니다:


In [7]:
table2=pd.DataFrame(index=[6,8,12]) # 주사위의 숫자

사전 확률과 우도를 나타내기 위해 분수를 사용하겠습니다. 이렇게 하면 실수로 반올림되지 않습니다.

In [8]:
from fractions import Fraction
table2['prior'] = Fraction(1,3)
table2['likelihood'] = Fraction(1,6), Fraction(1,8), Fraction(1,12)
table2

prior likelihood
6    1/3        1/6
8    1/3        1/8
12   1/3       1/12

사전 확률과 우도를 얻은 후, 나머지 단계는 항상 동일하므로 이를 함수로 정리하겠습니다:

In [13]:
table2['unnorm']=table2['prior']*table2['likelihood']

In [29]:
def update(table):
    """
        posterior(사후) 확률 계산
    """
#     table['unnorm']=table['prior']*table['likelihood']
    prob_data = table['unnorm'].sum()
    posterior = table['unnorm']/prob_data
    return posterior

In [15]:
table2['posterior'] = update(table2)
table2

prior likelihood unnorm posterior
6    1/3        1/6   1/18       4/9
8    1/3        1/8   1/24       1/3
12   1/3       1/12   1/36       2/9

6면 주사위의 사후 확률은 4/9로, 다른 주사위의 확률(3/9 및 2/9)보다 조금 높습니다. 직관적으로, 6면 주사위가 가장 가능성이 높은 이유는 우리가 관측한 결과를 생성할 가능성이 가장 높았기 때문입니다.

## Monty hall 문제
다음으로 베이스 테이블을 사용하여 확률 이론에서 가장 논란이 많은 문제 중 하나를 해결해 보겠습니다.

Monty Hall 문제는 "Let's Make a Deal"라는 게임 쇼를 기반으로 합니다. 만약 이 게임 쇼의 참가자라면 게임이 어떻게 진행되는지 살펴보겠습니다.

진행자 Monty Hall은 여러분에게 1, 2, 3으로 번호가 매겨진 세 개의 닫힌 문을 보여주고, 각 문 뒤에는 상품이 있다고 말합니다.

상품 중 하나는 가치가 있는 것(전통적으로 자동차)이고, 나머지 두 개는 가치가 적은 것(전통적으로 염소)입니다.

게임의 목적은 자동차가 있는 문을 추측하는 것입니다. 만약 올바르게 추측하면 자동차를 얻을 수 있습니다.

예를 들어, 당신이 문 1을 선택했다고 가정해 봅시다. 당신이 선택한 문을 열기 전에 Monty는 문 3을 열고 염소를 보여줍니다. 그런 다음 Monty는 당신에게 원래 선택을 유지하거나 남은 두 개의 닫힌 문 중 하나로 변경할 옵션을 제공합니다.


이 질문에 답하기 위해 진행자의 행동에 대한 몇 가지 가정을 해야 합니다:

1. Monty는 항상 문을 열고 당신에게 바꾸는 옵션을 제공합니다.

2. 그는 당신이 선택한 문이나 자동차가 있는 문을 열지 않습니다.

3. 만약 당신이 자동차가 있는 문을 선택한 경우, 그는 나머지 두 개의 문 중 하나를 무작위로 선택합니다.

이러한 가정 하에서는 문을 바꾸는 것이 더 좋습니다. 만약 원래 선택을 유지한다면 자동차를 얻는 확률은 1/3입니다. 그러나 문을 바꾸면 자동차를 얻는 확률은 2/3입니다.

이 문제를 이전에 접하지 않았다면, 이 답변이 놀라울 수 있습니다. 혼자가 아닙니다. 많은 사람들은 문을 바꿀지 유지할지는 중요하지 않다고 생각합니다. 두 개의 문이 남아 있으므로 자동차가 문 A 뒤에 있을 확률은 50%라고 이유할 것입니다. 그러나 이것은 틀린 주장입니다.

왜 그런지 이해하기 위해 베이스 테이블을 사용하는 것이 도움이 될 수 있습니다. 먼저 세 가지 가설로 시작합니다. 자동차는 문 1, 2 또는 3 뒤에 있을 수 있습니다. 문제의 진술에 따르면 각 문의 사전 확률은 1/3입니다.

In [21]:
table3=pd.DataFrame(index=['Door 1','Door 2','Door 3'])
table3['prior']=Fraction(1,3)
table3

prior
Door 1   1/3
Door 2   1/3
Door 3   1/3

데이터는 Monty가 문 3을 열고 염소를 보여준 사실입니다. 따라서 각 가설 하에서 데이터의 확률을 고려해 보겠습니다:

* 자동차가 문 1 뒤에 있다면, Monty는 문 2 또는 3 중 하나를 무작위로 선택하므로 문 3을 열 확률은 $1/2$입니다.

* 자동차가 문 2 뒤에 있다면, Monty는 문 3을 열어야 하므로 이 가설 하에서 데이터의 확률은 1입니다.

* 자동차가 문 3 뒤에 있다면, Monty는 이를 열지 않으므로 이 가설 하에서 데이터의 확률은 0입니다.

여기에 우도(likelihood)가 있습니다.

In [22]:
table3['likelihood']=Fraction(1,2),1,0
table3

prior likelihood
Door 1   1/3        1/2
Door 2   1/3          1
Door 3   1/3          0

이제 사전 확률과 우도를 가지고 있으므로 update를 사용하여 사후 확률을 계산할 수 있습니다.

Monty가 문 3을 열고 나서, 문 1의 사후 확률은 $1/3$이며, 문 2의 사후 확률은 $2/3$입니다. 따라서 Door 1에서 Door 2로 바꾸는 것이 유리합니다.

In [24]:
def update1(table7):
    """
        posterior(사후) 확률 계산
    """
    table7['unnorm']=table7['prior']*table7['likelihood']
    prob_data = table7['unnorm'].sum()
    table7['posterior'] = table7['unnorm']/prob_data
    return table7

In [25]:
update1(table3)

prior likelihood unnorm posterior
Door 1   1/3        1/2    1/6       1/3
Door 2   1/3          1    1/3       2/3
Door 3   1/3          0      0         0

이 장에서는 베이즈 정리를 명시적으로 사용하고 베이스 테이블을 사용하여 쿠키 문제를 해결했습니다. 이 두 방법 사이에 실제로 큰 차이는 없지만, 베이스 테이블은 특히 두 개 이상의 가설이 있는 문제의 총 데이터 확률을 계산하기 쉽게 만들어 줄 수 있습니다.

그런 다음 우리는 주사위 문제와 몬티 홀 문제를 해결했습니다. 몬티 홀 문제가 머리 아프게 만든다면, 혼자가 아닙니다. 그러나 이 문제는 까다로운 문제를 해결하기 위한 분할-정복 전략으로서 베이즈 정리의 힘을 보여주며, 왜 답이 그렇게 나오는지에 대한 통찰력을 제공할 것으로 기대합니다.

몬티가 문을 열면 자동차의 위치에 대한 우리의 믿음을 업데이트하는 데 사용할 수 있는 정보를 제공합니다. 이 정보의 일부는 명백합니다. 만약 그가 문 3을 열면, 자동차가 문 3 뒤에 없다는 것을 알게됩니다. 그러나 정보의 일부는 미묘합니다. 문 3을 열 확률은 자동차가 문 2 뒤에 있을 경우 더 높고, 문 1 뒤에 있을 경우 더 낮습니다. 따라서 데이터는 문 2를 지지하는 증거입니다. 이러한 증거의 개념에 대해 향후 장에서 논의하겠습니다.

다음 장에서는 쿠키 문제와 주사위 문제를 확장하고, 기본 확률에서 베이지안 통계로 나아가는 다음 단계를 살펴보겠습니다.

하지만 먼저 연습 문제를 해결해보는 것이 좋을 것입니다.

## 연습문제

상자안에 두개의 동전이 있다고 가정해봅시다. 하나는 앞면과 뒷면이 있는 보통 동전이고, 다른 하나는 양면이 앞면인 특별한 동전입니다. 여러분은 무작위로 동전을 선택하고 한 쪽면이 앞면인 것을 본다고 가정합니다. 여러분이 특별한 동전을 선택한 확률은 얼마일까요?

In [32]:
table4 = pd.DataFrame(index=['Normal','trick'])
table4['prior'] = 1/2
table4['likelihood']=1/2,1
update1(table4)

prior  likelihood  unnorm  posterior
Normal    0.5         0.5    0.25   0.333333
trick     0.5         1.0    0.50   0.666667

In [16]:
df = pd.DataFrame({
    'name':['홍길동','김개똥','김말자','국제','아카데미','홍1','홍2','홍3','홍4'],
    'ban':[1,3,1,1,3,2,4,2,3],
    'math':[98,23,56,11,78,88,99,100,11],
    'eng':[22,44,77,88,44,66,99,100,77]
})
df

name  ban  math  eng
0   홍길동    1    98   22
1   김개똥    3    23   44
2   김말자    1    56   77
3    국제    1    11   88
4  아카데미    3    78   44
5    홍1    2    88   66
6    홍2    4    99   99
7    홍3    2   100  100
8    홍4    3    11   77

In [18]:
def user_fn(x):
    result_dict={}
    try:
        result_dict['str_merge']= x.str.cat(sep=' ')
    except:
        result_dict = {i : ((1-i*0.2)*x.max()-(i*0.2)*x.min())/2 for i in range(7)}
        result_dict['sum']=sum(x)
    return result_dict

In [19]:
df_math_sum_and_cont = df.groupby('ban')['name'].agg([user_fn])
for i in df_math_sum_and_cont['user_fn']:
    for k,v in i.items():
        print(k,v)

str_merge 홍길동 김말자 국제
str_merge 홍1 홍3
str_merge 김개똥 아카데미 홍4
str_merge 홍2


In [20]:
df_math_mean=df.groupby('ban')['math'].mean()
df_math_sum_and_cont= df.groupby('ban')['math'].agg(['mean','sum','count'])
df_math_sum_and_cont['mean']==df_math_sum_and_cont['sum']/df_math_sum_and_cont['count']

ban
1    True
2    True
3    True
4    True
dtype: bool